### ISSUE TITLE
[Lua] SimulatedBody.GetOnCollisionBeginEvent() issue

### ISSUE THREAD

**Issue by yaakuro** [o3de/o3de#17630]
**Describe the bug**
When using Lua to detect collisions through the SimulatedBody.GetOnCollisionBeginEvent(...) function, it consistently returns nil. The reason for this issue can be found in the RigidBodyComponent.cpp file, where the following line exists:

```cpp
    void RigidBodyComponent::Activate()
    {
      ....
        // During activation all the collider components will create their physics shapes.
        // Delaying the creation of the rigid body to OnEntityActivated so all the shapes are ready.
        AZ::EntityBus::Handler::BusConnect(GetEntityId());
    }

    void RigidBodyComponent::OnEntityActivated([[maybe_unused]] const AZ::EntityId& entityId)
    {
        AZ::EntityBus::Handler::BusDisconnect();

        // Create and setup rigid body & associated bus handlers
        CreateRigidBody();

        // Add to world
        EnablePhysics();
    }
```

Upon examination, it appears that the creation of the rigid body is intentionally being deferred, as noted in the accompanying comment. However, this delay leads to a problem when attempting to detect collisions via Lua within the OnActivate() function.
```lua
function CollisionHandling:OnActivate()
    self.event = SimulatedBody.GetOnCollisionBeginEvent(self.entityId)
    self.CollisionBeginEvent = self.event:Connect(
        function(_, collision)
            Debug.Log("------>")
        end
    );
end
```
Not sure if this is my wrong doing or this is really a problem.

**Steps to reproduce**
1. Create Level
2. Create 2 entities with rigid body component
3. Add lua component and try to use the GetOnCollisionBeginEvent()
4. Start Game
5. See error

**Expected behavior**
SimulatedBody.GetOnCollisionBeginEvent(self.entityId) should not return nil.

**Actual behavior**
A clear and concise description of what actually happened.

**Screenshots/Video**
If applicable, add screenshots and/or a video to help explain your problem.

**Found in Branch**
SimulatedBody.GetOnCollisionBeginEvent(self.entityId) returns nil.

**Commit ID from [o3de/o3de](https://github.com/o3de/o3de) Repository**
26baf4a0e5904031469362641580047a61a2bd8a

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS:  Linux
 - Version Kubuntu 22.04.3
 - CPU Ryzen 5950X
 - GPU RX 6950 XT
 - Memory 128GB

**Comment by yaakuro** [comment]
I checked the BaseColliderComponent and to me it doesn't look like that the InitShapes needs to be done in the Activate method. It can be put into Init so the shapes will be ready when the components gets activated and the rigid bodies are created. It looks to me redundant to Init the shapes and clear everytime we activate and deactivate the component.

Also I was thinking, why would we want to create and destroy a rigid body ever time we activate/deactivate the component? Is there a reason it is as it is right now? Why not just create once then modify the instance and reset if needed.

**Comment by nick-l-o3de** [comment]
regarding activate/deactivate / Init...

The opposite of Activate, is Deactivate.  After deactivate, it can be reactivated again.
The opposite of Init is delete (destructor).   It cannot be salvaged after the destructor.

Its possible to add and remove components (including the rigid body component and collider A) while the entity is deactivated, but not while active.  This guarintees that until deactivate, all pointers to other components cached for performance remain valid.

Its also by convention when components are listening and active in the wold . Non-active components exist elsewhere, such as for example, when deserialized from file during compilation, or in the undo stack - they may be init, but not active.  Actually adding rigid bodies and shapes to the world during that phase (parallel compilation or "dead entities" in the undo stack) would be a super bad idea.  This is probably why shapes are destroyed during deactivate and created during activate.  Last thing you want is shapes actually being added to a physx world when its loading entities (deserialize + init, but not activate) in the Asset Processor to crunch them into spawnables.

That being said, it does sound like there's a sort of sequence of events here, to me it sounds like we are missing a callback of some sort here, so that things that depend on physics can hook into when physics is ready, still part of activate...?

**Comment by yaakuro** [comment]
Fair enough. I digged more into this and it looks like, that RigidBodyNotifications is used to handle initialization in using the events

```cpp
        virtual void OnPhysicsEnabled([[maybe_unused]] const AZ::EntityId& entityId)
        virtual void OnPhysicsDisabled([[maybe_unused]] const AZ::EntityId& entityId)
```

In short, in Lua it would like this:

```lua
function MyLuaScript:OnActivate()
        self.RigidBodyNotificationBusHandler = RigidBodyNotificationBus.Connect(self, self.entityId)
end

function MyLuaScript:OnDeactivate()
        self.RigidBodyNotificationBusHandler:Disconnect()
end
```

So looking into some examples like in ForceRegionComponent.cpp. It does this:

Activate: Connect to RigidBodyNotificationBus
Deactivate: Disconnect from RigidBodyNotificationBus

This however again, causes an issue on the Lua level because when disconnecting at OnDeactivate from the RigidBodyNotificationBus, the OnPhysicsDisabled event is not fired.

So using the new bus, RigidBodyNotificationBus I thought would help to "fix" this issue but then again, disconnecting on OnDeactivate seem to cause another unexpected issue.

**Comment by nick-l-o3de** [comment]
I'm not sure I 100% follow, I was expecting something like this ,with scopes not overlapping
```
  Entity Activated ---> lua invokes OnActivate
      Entity finishes setting up physics --> lua invokes OnPhysicsEnabled, you can now do stuff with physics
      .... game ticks here for a while
      ... time pases
   Entity is deactivated
        The physics are destroyed -- OnPhysicsDisabled is invoked
        Lua OnDeactivate is invoked.
```

I'm guessnig the order of that destruction is switched?
It shouldn't be - deactivation / destrction should be following the reverse order.  I can understand though, why someone might want to invoke the lua OnDeactivate at the beginning, before the entity is actually deactivated, so that all the systems will still in place, but I'm not sure that's idomatic when shutdown should be happening in reverse order to startup...

**Comment by yaakuro** [comment]
Here this is a simple Lua script. I do assume, connecting is done in the correct way, because I would assume, I do connect to RigidBodyNotificationBus in OnActivate, and disconnect in the OnDeactivate:

```lua
local CollisionHandlingTest = 
{
    Properties =
    {
    }  	
}

function CollisionHandlingTest:OnActivate()
    Debug.Log("OnActivate")

    -- Connect to the RigidBodyNotificationBus
    self.RigidBodyNotificationBusHandler = RigidBodyNotificationBus.Connect(self, self.entityId)
end

function CollisionHandlingTest:OnDeactivate()
    Debug.Log("OnDeactivate")

    -- Disconnect from the RigidBodyNotificationBus
    self.RigidBodyNotificationBusHandler:Disconnect()
end

function CollisionHandlingTest:OnPhysicsEnabled(entityId)
    Debug.Log("OnPhysicsEnabled")
end

function CollisionHandlingTest:OnPhysicsDisabled(entityId)
    Debug.Log("OnPhysicsDisabled")
end

return CollisionHandlingTest
```

This is a picture of the console:

![image](https://github.com/o3de/o3de/assets/7720708/96bfa3f7-4163-4026-9cfa-8b2cb4762481)

As you can see, the OnPhysicsDisabled event is not triggered because we do disconnect in OnDeactivate.  If I move the 
```self.RigidBodyNotificationBusHandler:Disconnect()``` into the OnPhysicsDisabled function itself, it will work

```lua
local CollisionHandlingTest = 
{
    Properties =
    {
    }  	
}

function CollisionHandlingTest:OnActivate()
    Debug.Log("OnActivate")

    -- Connect to the RigidBodyNotificationBus
    self.RigidBodyNotificationBusHandler = RigidBodyNotificationBus.Connect(self, self.entityId)
end

function CollisionHandlingTest:OnDeactivate()
    Debug.Log("OnDeactivate")


end

function CollisionHandlingTest:OnPhysicsEnabled(entityId)
    Debug.Log("OnPhysicsEnabled")
end

function CollisionHandlingTest:OnPhysicsDisabled(entityId)
    Debug.Log("OnPhysicsDisabled")

    -- Disconnect from the RigidBodyNotificationBus
    self.RigidBodyNotificationBusHandler:Disconnect()
end

return CollisionHandlingTest
```

Here the console:

![image](https://github.com/o3de/o3de/assets/7720708/647f708a-1800-4140-a829-641948663110)


So yes, the order is not good. I am not sure why I am the only one having this issue and If everyone is using Script Canvas, is it handled differently? 
In my point of view as a user I do expect buses to be connected in the OnActivate of the lua function, and disconnected in the OnDeactivate. Everything else is misleading and wrong architecture. Even though I tried to use the RigidBodyNOtificationBus, this also lead to wrong usecases. I think the Init, Activate and Deactivate is not enough to handle all usecases in O3DE. 
Also not having any documentation on this is not good for O3DE beginners. This is frustrating. I might make some tutorials on this bus, still this issue needs to be handled.
So any ideas how to handle (what do we need to change, fix etc.). Meaning, fix the order? Add more events where O3DE guarantees all busses are ready to be connected, disconencted etc.

**Comment by yaakuro** [comment]
Also moving the component up and down in the inspector but does not have any effect, because even though I put the lua script at the most top. when saving everything, restarting the editor, and looking at the order, the lua script is again at the bottom.
